File: lib/domain/fertilize/gateways/fertilize_gateway.rb
# frozen_string_literal: true

module Domain
  module Fertilize
    module Gateways
      class FertilizeGateway
        def find_by_id(fertilize_id)
          raise NotImplementedError, "Subclasses must implement find_by_id"
        end

        def create(create_input_dto)
          raise NotImplementedError, "Subclasses must implement create"
        end

        def update(fertilize_id, update_input_dto)
          raise NotImplementedError, "Subclasses must implement update"
        end

        # @param filter [Domain::Shared::ValueObjects::ReferenceIndexListFilter]
        def list_index_for_filter(filter)
          raise NotImplementedError, "Subclasses must implement list_index_for_filter"
        end

        def create_for_user(user, attrs)
          raise NotImplementedError, "Subclasses must implement create_for_user"
        end

        def update_for_user(user, id, attrs)
          raise NotImplementedError, "Subclasses must implement update_for_user"
        end

        def soft_delete_with_undo(user:, fertilize_id:, auto_hide_after:, translator:)
          raise NotImplementedError, "Subclasses must implement soft_delete_with_undo"
        end

        # 作成検証失敗時のスナップショット（永続化しない）
        # @return [Domain::Fertilize::Dtos::FertilizeMasterFormSnapshot]
        def fertilize_master_form_snapshot_after_create_failure!(user:, attributes:)
          raise NotImplementedError, "Subclasses must implement fertilize_master_form_snapshot_after_create_failure!"
        end

        # AI API: ユーザー肥料を agrr 応答の商品名で検索（なければ nil）。戻りは永続 Fertilize。
        def find_by_name(user_id:, name:)
          raise NotImplementedError, "Subclasses must implement find_by_name"
        end

      end
    end
  end
end

File: crates/agrr-domain/src/fertilize/gateways/fertilize_gateway.rs
use crate::fertilize::entities::FertilizeEntity;
use crate::shared::attr::AttrMap;
use crate::shared::dtos::Error;
use crate::shared::user::User;
use crate::shared::value_objects::reference_index_list_filter::ReferenceIndexListFilter;

/// Ruby: `Domain::Fertilize::Gateways::FertilizeGateway`
pub trait FertilizeGateway: Send + Sync {
    fn find_by_id(
        &self,
        fertilize_id: i64,
    ) -> Result<FertilizeEntity, Box<dyn std::error::Error + Send + Sync>>;

    fn list_index_for_filter(
        &self,
        filter: &ReferenceIndexListFilter,
    ) -> Result<Vec<FertilizeEntity>, Box<dyn std::error::Error + Send + Sync>>;

    fn create_for_user(
        &self,
        user: &User,
        attrs: AttrMap,
    ) -> Result<FertilizeEntity, Box<dyn std::error::Error + Send + Sync>>;

    fn update_for_user(
        &self,
        user: &User,
        id: i64,
        attrs: AttrMap,
    ) -> Result<FertilizeEntity, Box<dyn std::error::Error + Send + Sync>>;

    fn soft_delete_with_undo(
        &self,
        user: &User,
        fertilize_id: i64,
        auto_hide_after: i64,
        translator: &dyn crate::shared::ports::TranslatorPort,
    ) -> Result<SoftDeleteWithUndoOutcome, Box<dyn std::error::Error + Send + Sync>>;

    fn find_by_name(
        &self,
        user_id: i64,
        name: &str,
    ) -> Result<Option<FertilizeEntity>, Box<dyn std::error::Error + Send + Sync>>;
}

#[derive(Debug, Clone)]
pub enum SoftDeleteWithUndoOutcome {
    Success { undo: serde_json::Value },
    Failure(Error),
}
